### ISSUE TITLE
Include src in npm package

### ISSUE THREAD

**Issue by matthewp** [ramda/ramda#829]
It's common for developers using a commonjs loader to require only the package they are using. You would want to do something like

```js
var isEmpty = require('lamda/src/isEmpty');
```

This is especially important on the client where you don't want to bring in dependencies you're not using.

Additionally `src` is usually by convention a folder for files that then get built into the main package (if you were using an es6 transpiler for example).  I know that you are building to `dist` but if modules could be required directly `lib` might be a better place.

**Comment by CrossEye** [comment]
Funny, we just pulled it out.  But you're right, there is at least some good case for restoring it.

@buzzdecafe, @davidchambers: Any compelling reason to exclude it?  Anything more than tidiness?

**Comment by davidchambers** [comment]
> Funny, we just pulled it out.

We discussed the possibility in #811, but haven't actually done so.

**Comment by CrossEye** [comment]
Oh, I just assumed... :-)

Well maybe we should leave `src` in.  I do like this use-case.

**Comment by matthewp** [comment]
@CrossEye What about renaming it to `lib`?

**Comment by buzzdecafe** [comment]
i also like this use case, let's keep it

**Comment by CrossEye** [comment]
@matthewp: Can you make a compelling case for changing folder names?  It's not that I have a major objection, but I also don't want to do this without good reason.  Is there a standard, or an overwhelming consensus around `lib`?

**Comment by matthewp** [comment]
@CrossEye Sure, just that `src` usually denotes code that needs to be built before it can be used. `var isEmpty = require('ramda/src/isEmpty');` **feels** like an unsupported API even if it really is supported. `var isEmpty = require('ramda/lib/isEmpty');` still feels like you're doing something unintended but it should still work.  So my case is basically conventions, `src` is where es6/coffeescript kind of things go and `lib` is where you're library's javascript modules go.

What would be even better would be if you could do `var isEmpty = require('ramda/isEmpty');`.   Unfortunately Node doesn't provide a way to make this work while keeping your code somewhere other than the package's root folder...  You could do what Lodash does and have the npm-published code in a separate branch (he uses the `npm` branch) and not worry about code organization there.  The downside is that this requires an additional build step in your publishing process.

tldr; I'd be happy so long as we can directly require modules whether in src or lib or wherever.  If you did decide to go with option 3 (publishing from a different branch) i'd be happy to contribute there.

**Comment by davidchambers** [comment]
> What would be even better would be if you could do `var isEmpty = require('ramda/isEmpty');`. Unfortunately Node doesn't provide a way to make this work while keeping your code somewhere other than the package's root folder...

The tarball published via npm needn't have the same directory structure as the repository. If we're to officially support requiring individual modules we should make the experience as nice as possible. It shouldn't be too difficult to support both `require('ramda')` and `require('ramda/isEmpty')`.

**Comment by danielyogel** [comment]
@davidchambers. I think you're right about supporting both require('ramda') and require('ramda/isEmpty') style. Also, it would be a good idea to add this option to the README file, next to the Partial Builds part. 
I had to search in the issues to discover this option.

**Comment by davidchambers** [comment]
Good idea, @danielyogel. Let's make sure to update the README in the pull request which adds support for `require('ramda/isEmpty')`.

**Comment by danielyogel** [comment]
@davidchambers, I'll be glad to help if I can. If I haven't missed it -  " the pull request which adds support for require('ramda/isEmpty') " does not exist yet. I'll start be studying Ramda's source structure, when I'll have more free time, to see how I can help.

**Comment by davidchambers** [comment]
> If I haven't missed it - " the pull request which adds support for require('ramda/isEmpty') " does not exist yet.

Correct. Everyone is on board with the idea; it's now a question of approach. In broad terms I think we want the [prepublish script][1] to recursively copy the files in __src__ to the top level. We'd update __.gitignore__ to ignore these top-level files, and we'd update __.npmignore__ to ignore files in __src__. It's fiddly, but worth doing.


[1]: https://github.com/ramda/ramda/blob/v0.10.0/scripts/prepublish

**Comment by danielyogel** [comment]
Does the prepublish script has to be written using bash script ? Can I use just JavaScript ?

**Comment by davidchambers** [comment]
> Does the prepublish script has to be written using bash script ? Can I use just JavaScript ?

The script could be written in JavaScript, but `cp -r src/* .` is hard to beat. ;)

**Comment by CrossEye** [comment]
@davidchambers: I'd be more inclined to build a temporary folder structure to use for `npm publish`, copying things we want to it exactly as we like, publish from there, and then throw it away.  I'd rather we didn't try to tie it directly to our project layout.

@danielyogel: I would like some day to have our entire build be in JS.  I'd rather a custom simple Node script than Grunt or Gulp or Broccoli, but would take any of those over the combination we have now of make and Grunt.  It's just not something I want to spend any real effort on.

**Comment by matthewp** [comment]
How are you going to do the gitignore?  I don't know if you can say "ignore any file that is in the root directory that is named the same as one in src/".

**Comment by matthewp** [comment]
@CrossEye I think your idea is simplier.

```shell
mkdir publish && cd publish
cp -R ../src/* .
cp ../package.json .
```

**Comment by davidchambers** [comment]
That's a very nice approach, @matthewp, but [xyz][1] runs `npm publish` from the project root so we'll need to make some other changes if we do it this way.

> I'd rather we didn't try to tie it directly to our project layout.

The layout cannot change, since modules require other modules via relative paths. The only question is where the subtree is rooted.

To clarify, I'm not suggesting we move anything in the repository, just that we copy some files to new locations before running `npm publish`.

> How are you going to do the gitignore? I don't know if you can say "ignore any file that is in the root directory that is named the same as one in src/".

We could do something like this:

```
/*.js
!/Gruntfile.js
```


[1]: https://github.com/davidchambers/xyz/blob/v0.5.0/xyz#L165

**Comment by danielyogel** [comment]
@CrossEye @davidchambers, It was wondering if it's possible to use ES6 modules, `import {map, compose, filter} from 'ramda'`, and get only the specified functions in the browserify/webpack build ? I guess it's not possible technically.
I am no expert with bash scripts, but am using gulp for some time and it's very flexible and pragmatic. So @CrossEye, if you would like some day to have our entire build be in JS, I could try to help with that. 
I'll recommend gulp, because you can always choose to use [regular npm modules](https://medium.com/@sogko/gulp-browserify-the-gulp-y-way-bb359b3f9623) instead of gulp plugins (and warp them in a "gulp" stream - or not - as needed). For example, even the [official gulp recipes page](https://github.com/gulpjs/gulp/blob/master/docs/recipes/browserify-uglify-sourcemap.md) recommends to use the regular Browserify npm module. 
One problem with gulp, is that right now its tasks flow-control is very limited, but that's not a big problem - I just promisify the streams and use Bluebird for flow control. So actually, right now, **I use gulp - just as a bunch of npm modules with a convention for a compatible stream standard, that's all** (And using streams is objectively better, no matter if you use gulp - or not).
Also, [gulp 4](https://twitter.com/gulpjs) is to be released soon, and will provide better flow control.

**Comment by CrossEye** [comment]
@danielyogel I would definitely be interested in such a gulp-based build.

As to the build, it's easy to do a custom build that includes only the functions you want and their (recursive) dependencies.  See https://github.com/ramda/ramda#partial-builds

**Comment by buzzdecafe** [comment]
This appears to be working for some time:

```js
> map = require('ramda/src/map');
> map(function (x) {return x + 1; }, [1, 2, 3]);
[ 2, 3, 4 ]
```

**Comment by CrossEye** [comment]
The goal, though, is slightly different:

```
> map = require('ramda/map')
```
